drivers/ethernet/eth_stellaris.c

```diff
@@ -0,0 +1,338 @@
+				sys_write32(*(u32_t *)data_ptr, REG_MACDATA);
+			}
+		}
+	}
+
+	/* Write the last, partial word in net_buf */
+	if (bytes_left > 0) {
+		reg_val = 0;
+		switch (bytes_left) {
+		case 3:
+			reg_val |= ((u32_t)data_ptr[2] << 16);
+		case 2:
+			reg_val |= ((u32_t)data_ptr[1] << 8);
+		case 1:
+			reg_val |= (u32_t)data_ptr[0];
+			break;
+		default:
+			break;
+		}
+		sys_write32(reg_val, REG_MACDATA);
+	}
+
+	/* Enable Txn */
+	sys_write32(BIT_MACTR_NEWTX, REG_MACTR);
+
+	/* Wait and check if txn successfull or not */
+	k_sem_take(&dev_data->tx_sem, K_FOREVER);
+
+	if (dev_data->tx_err) {
+		SYS_LOG_ERR("Txn failed");
+		dev_data->tx_err = false;
+		return -EIO;
+	}
+
+	/* Update statistics counters */
+	eth_stats_update_bytes_tx(iface, net_pkt_get_len(pkt));
+	eth_stats_update_pkts_tx(iface);
+	pkt_hdr = NET_ETH_HDR(pkt);
+	if (net_eth_is_addr_multicast(&pkt_hdr->dst)) {
+		eth_stats_update_multicast_tx(iface);
+	} else if (net_eth_is_addr_broadcast(&pkt_hdr->dst)) {
+		eth_stats_update_broadcast_tx(iface);
+	}
+
+	SYS_LOG_DBG("pkt send &p len &d", pkt, net_pkt_get_len(pkt));
+	net_pkt_unref(pkt);
+
+	return 0;
+}
+
+void eth_stellaris_rx(struct device *dev)
+{
+	struct net_pkt *pkt = NULL;
+	struct eth_stellaris_runtime *dev_data = DEV_DATA(dev);
+	struct net_if *iface = dev_data->iface;
+	struct net_eth_hdr *pkt_hdr;
+	u32_t reg_val;
+	int pktlen, bytes_left, ret;
+
+	/* Obtain the packet to be populated */
+	pkt = net_pkt_get_reserve_rx(0, K_NO_WAIT);
+	if (!pkt) {
+		SYS_LOG_ERR("Could not allocate packet");
+		eth_stats_update_errors_rx(iface);
+		return;
+	}
+
+	/* Read the First word */
+	reg_val = sys_read32(REG_MACDATA);
+	pktlen = reg_val & 0x0000ffff;
+	if (!net_pkt_append(pkt, 2, (u8_t *)&reg_val + 2, K_NO_WAIT)) {
+		goto pkt_err;
+	}
+
+	/* Read the rest of words, minus the partial word and FCS byte */
+	for (bytes_left = pktlen - 4; bytes_left > 7; bytes_left -= 4) {
+		reg_val = sys_read32(REG_MACDATA);
+		if (!net_pkt_append(pkt, 4, (u8_t *)&reg_val, K_NO_WAIT)) {
+			goto pkt_err;
+		}
+	}
+
+	/* Handle the last partial word and Discard the 4 Byte FCS */
+	while (bytes_left > 0) {
+		reg_val = sys_read32(REG_MACDATA);
+		/* Discard the last FCS word */
+		if (bytes_left <= 4) {
+			bytes_left = 0;
+			break;
+		}
+		/* Read the the partial word */
+		if (!net_pkt_append(pkt, bytes_left - 4,
+				    (u8_t *)&reg_val, K_NO_WAIT)) {
+			goto pkt_err;
+		}
+		bytes_left -= 4;
+	}
+
+	/* Update statistics counters */
+	eth_stats_update_bytes_rx(iface, pktlen);
+	eth_stats_update_pkts_rx(iface);
+	pkt_hdr = NET_ETH_HDR(pkt);
+	if (net_eth_is_addr_broadcast(&pkt_hdr->dst)) {
+		eth_stats_update_broadcast_rx(iface);
+	} else if (net_eth_is_addr_multicast(&pkt_hdr->dst)) {
+		eth_stats_update_multicast_rx(iface);
+	}
+
+	ret = net_recv_data(iface, pkt);
+	if (ret < 0) {
+		goto frame_err;
+	}
+
+	goto done;
+
+frame_err:
+	SYS_LOG_ERR("Failed to place frame in RX queue:%d", ret);
```

Comment:
The fact that we are using variables here, that are set else where, means this should be closer to the place where the error is raised.